typescript/web/src/components/app-lifecycle-manager/welcome-modal/welcome-modal.tsx

```diff
@@ -1,185 +1,357 @@
-              maxW="lg"
-              color={mode("gray.600", "gray.300")}
-              fontWeight="extrabold"
-              textAlign="center"
-            >
-              The open standard{" "}
-              <Text color="brand.500" display="inline">
-                image labeling tool
-              </Text>
-            </Heading>
-
-            <Text
-              color={mode("gray.600", "gray.400")}
-              mt="16"
-              maxW="lg"
-              fontSize="lg"
-              fontWeight="medium"
-              textAlign="justify"
-            >
-              Stay in control of your data, label your images without them
-              leaving your computer. Focus on building the next big thing.
-            </Text>
-          </VStack>
-        </ModalBody>
-        <ModalFooter>
-          <HStack
-            direction={{ base: "column", md: "row" }}
-            justifyContent="center"
-            width="full"
-            spacing="4"
-            mb="10"
-          >
-            <Button
-              as="a"
-              leftIcon={<GithubIcon fontSize="xl" />}
-              href="https://github.com/Labelflow/labelflow"
-              target="blank"
-              size="lg"
-              minW="210px"
-              variant="link"
-              height="14"
-              px="8"
-            >
-              See code on Github
-            </Button>
-
-            <Button
-              ref={startLabellingButtonRef}
-              size="lg"
-              minW="210px"
-              colorScheme="brand"
-              height="14"
-              px="8"
-              isLoading={hasUserClickedStart && !isServiceWorkerActive}
-              onClick={handleClickStartLabelling}
-              loadingText="Loading the application"
-            >
-              Start Labelling!
-            </Button>
-          </HStack>
-        </ModalFooter>
-      </ModalContent>
+      {(() => {
+        // If there was an error AND the browser is wrong, we show the browser error modal
+        if (shouldShowBrowserErrorModal) {
+          return <BrowserError error={browserError} />;
+        }
+
+        // If there was an error, we handle it the normal way, showing the normal error screen
+        if (browserError) {
+          handleError(browserError);
+        }
+
+        // If there was no error AND the browser is wrong, we still show the browser warning modal
+        if (shouldShowBrowserWarningModal) {
+          return (
+            <BrowserWarning onClickTryAnyway={pretendIsCompatibleBrowser} />
+          );
+        }
+
+        // Nominal loading on first visit
+        if (shouldShowLoadingModal) {
+          return <Loading />;
+        }
+
+        // Nominal welcome modal on first vist after loading
+        if (shouldShowWelcomeModal) {
+          return (
+            <Welcome onClickSkip={handleSkip} onClickNext={handleGetStarted} />
+          );
+        }
+
+        return null;
+      })()}
```

Comment:
Nitpicking:

This could be a component instead of an IIFE.

Reply:
But then it would separate (in the code) this logic from the logic of the `isOpen` condition. See my previous answer about the condition. Both sides are related, so better to keep them together.

Reply:
You could define the child component right next to where you would define the isOpen const.
I know that this results in exactly the same thing from React's point of view. I think it's just more "react standard".

But that's fine like this if you prefer.